### ISSUE TITLE
gmpxx library

### ISSUE THREAD

**Issue by jeffreyhanson** [rwinlib/utils#1]
Hi,

Thank you very much for developing RWinLib! I was wondering if it would be possible to distribute the C++ bindings for the GMP library via RWinLib? This would be useful because (AFAIK) the version of GMP available on CRAN is a bit old and has issues when using the C++ bindings [(e.g. does not `mpq_cmp_z` based on logs from WinBUilder, per this log file)](https://github.com/rwinlib/utils/files/6660170/00install.out.txt). My impression is that the GMP build artefacts from  the https://github.com/r-windows/rtools-packages/tree/master/mingw-w64-gmp repository would be sufficient (since (1) they are built using the [`--enable-cxx` flag](https://github.com/r-windows/rtools-packages/blob/master/mingw-w64-gmp/PKGBUILD#L27) to provide the C++ bindings and (2) based on a fairly recent version of GMP). 

What do you think? If you don't mind creating a gmp or gmpxx repo under the RWinLib organization, I could try submiting a PR with the relevant files? I'm sorry if this isn't the right place to submit this issue. If you let me know the correct repo, I can close this issue and re-submit it there? Also, please let me know if there are any additionals details I can provide that might help?

**Comment by jeffreyhanson** [comment]
@jeroen (apologies if you're not the right person to contact about this), I just wanted to follow up and ask if it would be possible to add a GMPXX to RWinLib? I completely understand if you don't have the bandwidth for this at the moment.

**Comment by jeroen** [comment]
We can do that, but I think gmp and gmpxx are already included by default with rtools because they are a dependency of gcc. So if you are compiling an R package, you can just `-lgmpxx` and it should work.

**Comment by jeroen** [comment]
Ah I see now, the problem may be the winbuilder configuration who has an additional (older) gmp installed on the machine.

**Comment by jeroen** [comment]
OK I have added a copy here https://github.com/rwinlib/gmp

**Comment by jeffreyhanson** [comment]
> Ah I see now, the problem may be the winbuilder configuration who has an additional (older) gmp installed on the machine.

Yeah, I think that's the issue.

> OK I have added a copy here https://github.com/rwinlib/gmp

Wow - that's brilliant - thank you so much!! It's getting a bit late here, so I'll try it out tomorrow and let you it goes.

**Comment by jeroen** [comment]
Can you maybe first confirm if the problem has not already been solved on the winbuilder? I tried uploading the master branch of `surveyvoi` and it seems to work?

Btw, you can suppress a lot of noise in the build log by adding `PKG_CXXFLAGS = -Wno-ignored-attributes` to your `src/Makevars.win`

**Comment by jeffreyhanson** [comment]
Yeah, I can look into this and check if the issue is still present on WinBuilder - I'll get back to you later today. 

IIRC, I was able get survyevoi working on WinBuilder by [distributing a recent version of the `gmpxx.h` file with the package](https://github.com/jeffreyhanson/surveyvoi/tree/master/src/gmp), [using `Makevars.win` to pass a custom flag to detect when the package is being built on Windows](https://github.com/jeffreyhanson/surveyvoi/blob/master/src/Makevars.win#L2), and [manually linking to that `gmpxx.h` file when that custom flag is present (instead of system-wide installation of gmp)](https://github.com/jeffreyhanson/surveyvoi/blob/master/src/package.h#L30-L35). But this approach is really not ideal. I'll update the package to use https://github.com/rwinlib/gmp instead for Windows system (assuming that the issue still present on WinBuilder when building from it's installation of gmp).

Neat - thanks for the heads up - that's good to know!

**Comment by jeffreyhanson** [comment]
I've updated the main branch to use RWinLib for GMP dependencies and, from what I can tell, this seems to work (see [here for GitHub Actions build](https://github.com/jeffreyhanson/surveyvoi/actions/runs/1200184283)). Thanks again!

I've also verified that the GMP version on WinBuilder is still causing problems. Specifically, I created the [`norwinlib` branch](https://github.com/jeffreyhanson/surveyvoi/tree/norwinlib) by copying the main branch, and then removing/replacing the RWinLib components so that the package would use the version of GMP installed on the host system (see commits https://github.com/jeffreyhanson/surveyvoi/commit/7a959c3bfe0f1df31769bdb390cac972bef11c33, https://github.com/jeffreyhanson/surveyvoi/commit/22637a461bb99045879087066ea50d18c51f4e26). I then checked the `norwinlib` branch on WinBuilder (see [install log file](https://github.com/rwinlib/utils/files/7109230/00install.out.txt) and [check log file](https://github.com/rwinlib/utils/files/7109229/00check.log.txt), and still throws the following error (among others):

```
D:/Compiler/rtools40/mingw64/include/gmpxx.h:893:12: error: 'mpq_cmp_z' was not declared in this scope
```

What do you think? Please let me know if there's anything else I can do to help? 

EDIT: I didn't include `PKG_CXXFLAGS = -Wno-ignored-attributes` in the Makevars because this threw a warning (during package checks) about non-portable package flags - but I can rerun the WinBuilder checks with this flag included if you want?

**Comment by jeroen** [comment]
We can try to ask the winbuilder sysadmin, Uwe, if he can remove the old version.

**Comment by jeffreyhanson** [comment]
Ah ok, sounds good - thank you. Do you know if removing the old version from WinBuilder would also guruantee that a similar issue would not be encountered when running checks on CRAN's systems? In other words, if the old version is removed from WinBuilder, is it possible that the Windows package checks might pass on WinBuilder but fail on CRAN (because CRAN might have an older version of GMP installed too)? Or are CRAN's Windows package checks run on exactly the same system as WinBuilder?

**Comment by jeroen** [comment]
CRAN == winbuilder. There is only one windows machine, which is both used for testing and building the official binaries.

**Comment by jeffreyhanson** [comment]
Ah, I see - thanks!
